program/rust/src/tests/pyth_simulator.rs

```diff
@@ -34,41 +41,112 @@ use crate::c_oracle_header::{
+            panic!("Unable to locate {}", "pyth_oracle.so");
+        }));
+
+
+        let mut program_test = ProgramTest::default();
+        let program_key = Pubkey::new_unique();
+        let programdata_key = Pubkey::new_unique();
+
+        let upgrade_authority_keypair = Keypair::new();
+
+        let program_deserialized = UpgradeableLoaderState::Program {
+            programdata_address: programdata_key,
+        };
+        let programdata_deserialized = UpgradeableLoaderState::ProgramData {
+            slot:                      1,
+            upgrade_authority_address: Some(upgrade_authority_keypair.pubkey()),
+        };
+
+        // Program contains a pointer to progradata
+        let program_vec = bincode::serialize(&program_deserialized).unwrap();
+        // Programdata contains a header and the binary of the program
+        let mut programdata_vec = bincode::serialize(&programdata_deserialized).unwrap();
+        programdata_vec.append(&mut bpf_data);
+
+        let program_account = Account {
+            lamports:   Rent::default().minimum_balance(program_vec.len()),
+            data:       program_vec,
+            owner:      bpf_loader_upgradeable::ID,
+            executable: true,
+            rent_epoch: Epoch::default(),
+        };
+        let programdata_account = Account {
+            lamports:   Rent::default().minimum_balance(programdata_vec.len()),
+            data:       programdata_vec,
+            owner:      bpf_loader_upgradeable::ID,
+            executable: false,
+            rent_epoch: Epoch::default(),
+        };
+
+        // Add to both accounts to program test, now the program is deploy as upgradable
+        program_test.add_account(program_key, program_account);
+        program_test.add_account(programdata_key, programdata_account);
+
+
+        // Start validator
+        let (banks_client, payer, recent_blockhash) = program_test.start().await;
+
+        // Hack to duplicate a keypair
+        let genesis_keypair = Keypair::from_bytes(&payer.to_bytes()).unwrap();
+
+        let mut result = PythSimulator {
+            program_id: program_key,
             banks_client,
             payer,
             last_blockhash: recent_blockhash,
-        }
+            programdata_id: programdata_key,
+            upgrade_authority: upgrade_authority_keypair,
+            genesis_keypair,
+        };
+
+        // Transfer money to upgrade_authority so it can call the instructions
+        result
+            .airdrop(&result.upgrade_authority.pubkey(), 1000 * LAMPORTS_PER_SOL)
+            .await
+            .unwrap();
+
+        return result;
+    }
+
+
+    pub fn set_generic_as_payer(&mut self) {
+        self.payer = Keypair::from_bytes(&self.genesis_keypair.to_bytes()).unwrap();
+    }
+
+    pub fn set_upgrade_authority_as_payer(&mut self) {
+        self.payer = Keypair::from_bytes(&self.upgrade_authority.to_bytes()).unwrap();
```

Comment:
I suggest passing `payer` as an argument to `upd_permissions` instead of doing this. it'll be easier to understand and it's more flexible.

Reply:
Meh the reason I ended up doing this is because I liked everything being private inside of pyth_simulator

Reply:
Done